File: leetcode/algorithms/283_move_zeroes/main.java
class MoveZeroes {
    /**
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(N)
     */
    public void moveZeroes(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = 1; j < nums.length - i; j++) {
                if (nums[j - 1] == 0) {
                    int temp = nums[j - 1];
                    nums[j - 1] = nums[j];
                    nums[j] = temp;
                }
            }
        }
    }


    // Solution
    /**
     * Solution 1
     *
     * Two Pointers Swap
     *
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public void solution1(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return;
        }

        int lastNonZeroFoundAt = 0;

        for (int cur = 0; cur < nums.length; cur++) {
            if (nums[cur] != 0) {
                if (cur != lastNonZeroFoundAt) {
                    int temp = nums[lastNonZeroFoundAt];
                    nums[lastNonZeroFoundAt] = nums[cur];
                    nums[cur] = temp;
                }

                lastNonZeroFoundAt++;
            }
        }
    }

    /**
     * Solution 2
     *
     * Overwrite & Fill
     *
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public void solution2(int[] nums) {
        int insertPos = 0;

        for (int num : nums) {
            if (num != 0) {
                nums[insertPos++] = num;
            }
        }

        while (insertPos < nums.length) {
            nums[insertPos++] = 0;
        }
    }
}

File: leetcode/algorithms/283_move_zeroes/main.cpp
#include <algorithm> // swap, stable_partition
#include <vector>
using namespace std;

class MoveZeroes {
public:
    /**
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(N)
     */
    void moveZeroes(vector<int>& nums) {
        int newNums[nums.size()];
        int count = 0;

        for (int i = 0; i < nums.size(); i++) {
            if (nums[i]) {
                newNums[count++] = nums[i];
            }
        }

        for (int i = 0; i < count; i++) {
            nums[i] = newNums[i];
        }

        for (int i = 1; i <= nums.size() - count; i++) {
            nums[nums.size() - i] = 0;
        }
    }


    // Solution
    /**
     * Solution 1
     * 
     * std::swap
     *
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    void solution1(vector<int>& nums) {
        int lastNonZeroFoundAt = 0;

        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] != 0) {
                if (i != lastNonZeroFoundAt) {
                    swap(nums[lastNonZeroFoundAt], nums[i]);
                }
                lastNonZeroFoundAt++;
            }
        }
    }

    /**
     * Solution 2
     * 
     * std::stable_partition
     *
     * Complexities:
     *   N - The size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    void solution2(vector<int>& nums) {
        stable_partition(nums.begin(), nums.end(), [](int n) {
            return n != 0;
        });
    }
};
